services/cloudapi/cloudapiserver/server.go

```diff
@@ -298,33 +347,102 @@ func (ss *Service) Stop(ctx context.Context, datacenterID, serverID, serverType
 		apiResponse.LogInfo()
 		if err != nil {
 			return err
 		}
-		return utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVmStateFn(ctx, datacenterID, constant.CubeVMStateStop))
+		if err = utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVMStateFn(datacenterID, constant.CubeVMStateStop)); err != nil {
+			return err
+		}
+		if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+			return fmt.Errorf("an error occurred while waiting for server state change on VM SUSPEND dcId: %s, server_id: %s, Response: (%w)", datacenterID, serverID, errState)
+		}
+		log.Printf("[DEBUG] %s server suspended: serverId: %s \n", serverType, serverID)
+		return nil
 	}
 
 	return fmt.Errorf("cannot stop unknown server type: %s", serverType)
 
 }
 
-// checkExpectedVmStateFn wraps over the ResourceReadyFunc to allow passing expectedState
-// TODO: change ResourceReadyFunc sig to support passing an expectedState param
-func (ss *Service) checkExpectedVmStateFn(ctx context.Context, dcId, expectedState string) utils.ResourceReadyFunc {
+// Reboot resets the VM,
+func (ss *Service) Reboot(ctx context.Context, datacenterID, serverID string) error {
+
+	apiResponse, err := ss.Client.ServersApi.DatacentersServersRebootPost(ctx, datacenterID, serverID).Execute()
+	apiResponse.LogInfo()
+	if err != nil {
+		return err
+	}
+	if err = utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVMStateFn(datacenterID, constant.VMStateStart)); err != nil {
+		return err
+	}
+	if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+		return fmt.Errorf("an error occurred while waiting for server state change on reboot dcId: %s, server_id: %s, Response: (%w)", datacenterID, serverID, errState)
+	}
+	log.Printf("[DEBUG] server reboot finished: serverId: %s \n", serverID)
+	return nil
+}
+
+// PxeBoot restarts the Server into the PXE shell by unsetting the primary boot device of the VM and performing a reset
+func (ss *Service) PxeBoot(ctx context.Context, datacenterID, serverID string) error {
+
+	deviceID, deviceType, err := ss.GetCurrentBootDeviceID(ctx, datacenterID, serverID)
+	if err != nil {
+		if errors.Is(err, ErrNoBootDevice) {
+			return ss.Reboot(ctx, datacenterID, serverID)
+		}
+		return err
+	}
+
+	switch deviceType {
+	case constant.BootDeviceTypeVolume:
+		volumeProperties := ionoscloud.VolumeProperties{}
+		volumeProperties.SetBootOrder(constant.VolumeBootOrderNone)
+		_, apiResponse, err := ss.Client.VolumesApi.DatacentersVolumesPatch(ctx, datacenterID, deviceID).Volume(volumeProperties).Execute()
+		if err != nil {
+			return err
+		}
+		if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+			return errState
+		}
+		log.Printf("[DEBUG] enabled PXE boot for server: serverId: %s \n", serverID)
```

Comment:
Maybe should be more similar to the debug below where we explain what we did to the cdrom image?

Reply:
also added the id of the volume that was unset before rebooting ([555645e](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/521/commits/555645efa0ea6f7cdbb227cfb1aae3fb448e341d))